File: 5. Estructuras de datos no lineales/Python/arboles_avl.py
# 1. Implementa un árbol AVL que permita insertar nodos y mantenga el balance automáticamente.
# 2. Escribe una función para imprimir los nodos del árbol AVL por niveles.


class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.izquierda = None
        self.derecha = None
        self.altura = 1


class ArbolAVL:
    def insertar(self, raiz, valor):
        if not raiz:
            return Nodo(valor)
        elif valor < raiz.valor:
            raiz.izquierda = self.insertar(raiz.izquierda, valor)
        else:
            raiz.derecha = self.insertar(raiz.derecha, valor)

        raiz.altura = 1 + max(
            self.obtener_altura(raiz.izquierda), self.obtener_altura(raiz.derecha)
        )

        balance = self.obtener_balance(raiz)

        # Caso 1 - Izquierda
        if balance > 1 and valor < raiz.izquierda.valor:
            return self.rotar_derecha(raiz)

        # Caso 2 - Derecha
        if balance < -1 and valor > raiz.derecha.valor:
            return self.rotar_izquierda(raiz)

        # Caso 3 - Izquierda Derecha
        if balance > 1 and valor > raiz.izquierda.valor:
            raiz.izquierda = self.rotar_izquierda(raiz.izquierda)
            return self.rotar_derecha(raiz)

        # Caso 4 - Derecha Izquierda
        if balance < -1 and valor < raiz.derecha.valor:
            raiz.derecha = self.rotar_derecha(raiz.derecha)
            return self.rotar_izquierda(raiz)

        return raiz

    def obtener_altura(self, raiz):
        if not raiz:
            return 0
        return raiz.altura

    def obtener_balance(self, raiz):
        if not raiz:
            return 0
        return self.obtener_altura(raiz.izquierda) - self.obtener_altura(raiz.derecha)

    def rotar_derecha(self, z):
        y = z.izquierda
        T3 = y.derecha

        y.derecha = z
        z.izquierda = T3

        z.altura = 1 + max(
            self.obtener_altura(z.izquierda), self.obtener_altura(z.derecha)
        )
        y.altura = 1 + max(
            self.obtener_altura(y.izquierda), self.obtener_altura(y.derecha)
        )

        return y

    def rotar_izquierda(self, z):
        y = z.derecha
        T2 = y.izquierda

        y.izquierda = z
        z.derecha = T2

        z.altura = 1 + max(
            self.obtener_altura(z.izquierda), self.obtener_altura(z.derecha)
        )
        y.altura = 1 + max(
            self.obtener_altura(y.izquierda), self.obtener_altura(y.derecha)
        )

        return y

    def imprimir_niveles(self, raiz):
        if not raiz:
            return

        cola = []
        cola.append(raiz)

        while len(cola) > 0:
            print(cola[0].valor, end=" ")
            nodo = cola.pop(0)

            if nodo.izquierda:
                cola.append(nodo.izquierda)
            if nodo.derecha:
                cola.append(nodo.derecha)


arbol = ArbolAVL()
raiz = None
raiz = arbol.insertar(raiz, 10)
raiz = arbol.insertar(raiz, 20)
raiz = arbol.insertar(raiz, 30)
raiz = arbol.insertar(raiz, 40)
raiz = arbol.insertar(raiz, 50)
raiz = arbol.insertar(raiz, 25)
raiz = arbol.insertar(raiz, 35)
raiz = arbol.insertar(raiz, 45)
raiz = arbol.insertar(raiz, 55)
arbol.imprimir_niveles(raiz)

File: 5. Estructuras de datos no lineales/JavaScript/arboles_avl.js
// 1. Implementa las rotaciones necesarias para un árbol AVL y realiza pruebas de inserción de nodos.
// 2. Escribe una función para calcular el factor de balance de cada nodo en un árbol AVL.

// Implementación de un árbol AVL
class Nodo {
  constructor(valor) {
    this.valor = valor;
    this.izquierda = null;
    this.derecha = null;
    this.altura = 1;
  }
}

class AVL {
  constructor() {
    this.raiz = null;
  }

  // Función para obtener la altura de un nodo
  altura(nodo) {
    if (nodo === null) return 0;
    return nodo.altura;
  }

  // Función para obtener el máximo de dos números
  maximo(a, b) {
    return a > b ? a : b;
  }

  // Función para obtener el factor de balance de un nodo
  factorBalance(nodo) {
    if (nodo === null) return 0;
    return this.altura(nodo.izquierda) - this.altura(nodo.derecha);
  }

  // Rotación simple a la derecha
  rotacionDerecha(nodo) {
    let aux = nodo.izquierda;
    nodo.izquierda = aux.derecha;
    aux.derecha = nodo;
    nodo.altura =
      this.maximo(this.altura(nodo.izquierda), this.altura(nodo.derecha)) + 1;
    aux.altura =
      this.maximo(this.altura(aux.izquierda), this.altura(aux.derecha)) + 1;
    return aux;
  }

  // Rotación simple a la izquierda

  rotacionIzquierda(nodo) {
    let aux = nodo.derecha;
    nodo.derecha = aux.izquierda;
    aux.izquierda = nodo;
    nodo.altura =
      this.maximo(this.altura(nodo.izquierda), this.altura(nodo.derecha)) + 1;
    aux.altura =
      this.maximo(this.altura(aux.izquierda), this.altura(aux.derecha)) + 1;
    return aux;
  }

  // Rotación doble a la derecha
  rotacionDobleDerecha(nodo) {
    nodo.izquierda = this.rotacionIzquierda(nodo.izquierda);
    return this.rotacionDerecha(nodo);
  }

  // Rotación doble a la izquierda

  rotacionDobleIzquierda(nodo) {
    nodo.derecha = this.rotacionDerecha(nodo.derecha);
    return this.rotacionIzquierda(nodo);
  }

  // Función para insertar un nodo en el árbol AVL
  insertar(valor) {
    this.raiz = this.insertarNodo(this.raiz, valor);
  }

  insertarNodo(nodo, valor) {
    if (nodo === null) {
      return new Nodo(valor);
    }
    if (valor < nodo.valor) {
      nodo.izquierda = this.insertarNodo(nodo.izquierda, valor);
    } else if (valor > nodo.valor) {
      nodo.derecha = this.insertarNodo(nodo.derecha, valor);
    } else {
      return nodo;
    }
    nodo.altura =
      1 + this.maximo(this.altura(nodo.izquierda), this.altura(nodo.derecha));
    let factor = this.factorBalance(nodo);
    if (factor > 1 && valor < nodo.izquierda.valor) {
      return this.rotacionDerecha(nodo);
    }
    if (factor < -1 && valor > nodo.derecha.valor) {
      return this.rotacionIzquierda(nodo);
    }
    if (factor > 1 && valor > nodo.izquierda.valor) {
      return this.rotacionDobleDerecha(nodo);
    }
    if (factor < -1 && valor < nodo.derecha.valor) {
      return this.rotacionDobleIzquierda(nodo);
    }
    return nodo;
  }

  // Función para recorrer el árbol en orden
  inOrden(nodo) {
    if (nodo !== null) {
      this.inOrden(nodo.izquierda);
      console.log(nodo.valor);
      this.inOrden(nodo.derecha);
    }
  }
}

// Prueba de inserción de nodos en un árbol AVL
let arbol = new AVL();
arbol.insertar(10);
arbol.insertar(20);
arbol.insertar(30);
arbol.insertar(40);
arbol.insertar(50);
arbol.insertar(25);
arbol.inOrden(arbol.raiz);

// Prueba de cálculo del factor de balance de cada nodo en un árbol AVL
console.log(arbol.factorBalance(arbol.raiz));
console.log(arbol.factorBalance(arbol.raiz.izquierda));
console.log(arbol.factorBalance(arbol.raiz.derecha));
console.log(arbol.factorBalance(arbol.raiz.derecha.derecha));
console.log(arbol.factorBalance(arbol.raiz.derecha.izquierda));
